torch/csrc/jit/codegen/cuda/parallel_dimension_map.cpp

```diff
@@ -0,0 +1,296 @@
+      exact_types_.insert(pt);
+    } else {
+      // Multiple constant dimensions found; Use the corresponding
+      // symbolic parallel dim
+      dim_map_.insert({pt, kir::NamedScalar::getParallelDim(pt)});
+    }
+  } else {
+    // Prefer to use blockDim/gridDim if not constant
+    dim_map_.insert({pt, kir::NamedScalar::getParallelDim(pt)});
+    exact_types_.insert(pt);
+  }
+}
+
+void ParallelDimensionMap::populateDimensionMapWithMultipleCASet(
+    ParallelType pt,
+    const std::unordered_set<IterDomain*>& dom_set) {
+  TORCH_INTERNAL_ASSERT(dom_set.size() > 1);
+
+  const auto gpu_lower = GpuLower::current();
+  kir::IrBuilder ir_builder(gpu_lower->kernel());
+
+  bool all_equal = true;
+  kir::Val* known_dimension =
+      gpu_lower->lowerValue((*dom_set.begin())->extent());
+  // Set it -1 to signal it's not initialied yet
+  int64_t known_const = -1;
+
+  // Check all of concrete domains to see if they match all together.
+  for (auto concrete_id : dom_set) {
+    // If this concrete domain has a constant extent, check if it
+    // matches with the known constant extent.
+    auto it = constant_extent_map_.find(concrete_id);
+    if (it != constant_extent_map_.end()) {
+      const auto& const_extent_set = it->second;
+      // If multiple constants are detected, it's not exact.
+      if (const_extent_set.size() > 1) {
+        all_equal = false;
+        break;
+      }
+      auto this_const = *(const_extent_set.begin());
+      // known_const is initialized to -1
+      if (known_const == -1) {
+        known_const = this_const;
+      } else if (known_const == this_const) {
+        // Matched with previously known const. The extent of this
+        // domain must be equal to that's previously known.
+        continue;
+      } else {
+        // Unmatched. This dom_set extents may not be unique.
+        all_equal = false;
+        break;
+      }
+    }
+
+    // At this point, it still remains undetermined whether this id
+    // matches with those previously looked at. Constant check failed,
+    // but symbolic matching may succeed.
+    if (!equalDim(
+            known_dimension, gpu_lower->lowerValue(concrete_id->extent()))) {
+      all_equal = false;
+      break;
+    }
+  }
+
+  // If all_equal is still true, the dimension of this paralel type
+  // must be exact.
+  if (all_equal) {
+    exact_types_.insert(pt);
+  }
+  // Use the const value, if found, as its dimension
+  if (all_equal && known_const != -1) {
+    dim_map_.insert({pt, ir_builder.create<kir::Int>(known_const)});
+  } else {
+    dim_map_.insert({pt, kir::NamedScalar::getParallelDim(pt)});
+  }
+}
+
+kir::Val* ParallelDimensionMap::get(ParallelType pt) const {
+  TORCH_INTERNAL_ASSERT(isParallelTypeThread(pt), "Invalid ParallelType: ", pt);
+  auto it = dim_map_.find(pt);
+  if (it == dim_map_.end()) {
+    return nullptr;
+  } else {
+    return it->second;
+  }
+}
+
+bool ParallelDimensionMap::isExact(ParallelType pt) const {
```

Comment:
Is this method used for testing only or there will be other places that will rely on it?

Reply:
I intend to use it to fix #1052. Haven't tried it yet.